### ISSUE TITLE
Whisper Decoder Batch Prompting Issue

### ISSUE THREAD

**Issue by SaadKaleem** [NVIDIA/TensorRT-LLM#1756]
Hello,

I'd like to transcribe multiple different audios in a batched manner, with varying context lengths. The official example is not reflective of this, as it just [repeats](https://github.com/NVIDIA/TensorRT-LLM/blob/9bd15f1937f52658fb116f30d58fea786ce5d03b/examples/whisper/run.py#L266) the batched prompt. 

Therefore I've done the following:

* Audios were passed in the range of `1` second to `30` seconds chunks, for every batch, with zero sample padding done to the right, on each batch dimension to get `480000` frames (`30` seconds at `16kHz`)
* Mel features were computed, 3D Tensor of shape: `[batch_size (16), n_mels (128), 3000]`, which is passed into the encoder, providing an output of shape: `[batch_size (16), max_encoder_input_length (1500) , hidden_size (1280)]`

Because of varying initial prompts across the different batches, padding is applied to the context_length with `<|endoftext|>`.

* `decoder_input_ids` is 2D Tensor, of shape: `[batch_size, max_context_length_padded)]`
* As an example, the engine was built with `max_input_len` and `max_output_len` both being: `1024`
* I've ensured that `max_context_length` + `max_new_tokens` are well below `1024`.  (I think this is the upper limit?)

Example batch prompt:

```
<|startofprev|> prompt text here ... <|startoftranscript|><|es|><|transcribe|><|notimestamps|><|endoftext|>
<|startoftranscript|><|es|><|transcribe|><|notimestamps|><|endoftext|><|endoftext|><|endoftext|>
<|startoftranscript|><|es|><|transcribe|><|notimestamps|><|endoftext|><|endoftext|><|endoftext|>
<|startofprev|> prompt text here ... <|startoftranscript|><|es|><|transcribe|><|notimestamps|><|endoftext|>
...

```
Example way of batching the prompts:

```python
# sot_sequence is simply, encoded `<|startoftranscript|><|es|><|transcribe|><|notimestamps|>`

prompt_ids = [
  (
      torch.tensor(
          [self.tokenizer.sot_prev]
          + self.tokenizer.encode(text=initial_prompt, allowed_special=self.tokenizer.special_tokens_set)
          + self.tokenizer.sot_sequence
      )
      if initial_prompt != ""
      else torch.tensor(self.tokenizer.sot_sequence)
  )
  for initial_prompt in initial_prompts
]

decoder_input_ids = pad_sequence(
    prompt_ids, batch_first=True, padding_value=self.eot_id
)
```
### Environment

* CPU architecture: x86_64
* CPU/Host memory size: AMD EPYC 7H12 64-Core Processor - 196Gi RAM
* GPU properties
  * GPU name: NVIDIA A100-PCIE
  * GPU memory size: 40GB
* Libraries
  * TensorRT-LLM tag v0.9.0
  * CUDA version: release 12.2, V12.2.140 - Build cuda_12.2.r12.2/compiler.33191640_0
  * NVIDIA driver version: 535.129.03 
  * OS: Ubuntu 22.04.3 LTS

### Expected behavior 

To get an output of `max_seq_length` with transcriptions after `<|notimestamps|>`:

```
<|startofprev|> prompt text here ... <|startoftranscript|><|es|><|transcribe|><|notimestamps|> transcription here ... <|endoftext|>
<|startoftranscript|><|es|><|transcribe|><|notimestamps|> transcription ... <|endoftext|><|endoftext|><|endoftext|>
<|startoftranscript|><|es|><|transcribe|><|notimestamps|> transcription ... <|endoftext|><|endoftext|><|endoftext|>
<|startofprev|> prompt text here ... <|startoftranscript|><|es|><|transcribe|><|notimestamps|> transcription ... <|endoftext|>
```

---

### Actual Behavior

I'm not suppressing any tokens

What I'm getting are non-deterministic outputs of  `max_seq_length`, however with mostly:

* Empty transcriptions:
```
<|startoftranscript|><|es|><|transcribe|><|notimestamps|><|endoftext|><|endoftext|><|endoftext|> 
<|startofprev|> prompt text here ...<|startoftranscript|><|es|><|transcribe|><|notimestamps|><|endoftext|><|endoftext|><|endoftext|> 
``` 

* Strange behavior where the sequence `<|startoftranscript|><|es|><|transcribe|><|notimestamps|>` is repeating after <|endoftext|>

```
 <|startofprev|>some prompt here <|startoftranscript|><|es|><|transcribe|><|notimestamps|><|endoftext|><|startoftranscript|><|es|><|transcribe|><|notimestamps|> fuertes pendientes<|endoftext|>...
```

* Repeated words (hallucinations?) not ending with `<|endoftext|>`:
```
<|startoftranscript|><|es|><|transcribe|><|notimestamps|><|endoftext|><|endoftext|><|endoftext|> y y y y y y y y
<|startofprev|> prompt text here ... <|startoftranscript|><|es|><|transcribe|><|notimestamps|><|endoftext|><|endoftext|><|endoftext|> y y y y y y y y
``` 

---

### Observations

* Everything seems fine if no prompt in the batch has `<|startofprev|>` - thus, no padding being applied to the context input.
  * This seems to occur when mixing batches of prompts starting with `<|startofprev|>`, with other prompts starting without a previous context `<|startoftranscript|>`
* Given some previous context, and audio chunks - results from the decoder are non-deterministic across multiple runs. Keeping `beam_size`, `temperature`, etc, sampling config constant.
* Regarding padding:
  *  #1081 `--remove_input_padding` option is False, as Whisper doesn't support it.
  * For context padding, since I'm using the same `pad_id` as `eot_id`, it is possible that this has caused some context dominance in the attention scores, leading to outputting the same `pad_id` tokens.  
* `gpt_attention_plugin` (float16) and `cross_attention` are True for `ModelConfig`
   * disabling `gpt_attention_plugin` significantly increases latency times. 
* It [seems](https://github.com/NVIDIA/TensorRT-LLM/blob/250d9c293d5edbc2a45c20775b3150b1eb68b364/tensorrt_llm/runtime/generation.py#L1851) `attention_mask` is not passed to the decoder model during `use_gpt_attention_plugin`. I don't know the details behind this.
  * `position_ids` and `last_token_ids` are returned - however, it doesn't seem `pad_id` is considered in any way for the context inputs?

---

Could anyone help as to what I'm possibly doing wrong?

**Comment by SaadKaleem** [comment]
Update: 

The issue certainly was context padding, as adding the padding before the `<|startoftranscript|>` instead has helped improve the overall performance and resolved the subtle bugs. I'm currently verifying if the WER is up to par, and will close if it is.

**Comment by guialfaroben** [comment]
@SaadKaleem I am having the same issues. How did you come around to fix it? By removing the padding you mean that no prompting is possible?

**Comment by guialfaro053** [comment]
@SaadKaleem I am having the same issues. How did you come around to fix it? By removing padding before the start of transcription token you mean no prompting is possible?

**Comment by SaadKaleem** [comment]
> @SaadKaleem I am having the same issues. How did you come around to fix it? By removing padding before the start of transcription token you mean no prompting is possible?

No, infact by _adding_ the padding before `<|startoftranscript|>` seems to fix it.
